source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c

```diff
@@ -0,0 +1,304 @@
+    //
+    pInitCfg->rxBuffer = Ethernet_rxBuffer;
+
+    //
+    // The Application handle is not used by this application
+    // Hence using a dummy value of 1
+    //
+    Ethernet_getHandle((Ethernet_Handle)1, pInitCfg , &emac_handle);
+
+    //
+    //Do global Interrupt Enable
+    //
+    (void)Interrupt_enableInProcessor();
+
+    //
+    //Assign default ISRs
+    //
+    Interrupt_registerHandler(INT_EMAC_TX0, Ethernet_transmitISR);
+    Interrupt_registerHandler(INT_EMAC_RX0, Ethernet_receiveISR);
+//    Interrupt_registerHandler(INT_EMAC, Ethernet_genericISR);
+
+    //
+    //Change the priority of the interrupt handlers
+    //
+
+    Interrupt_setPriority(INT_EMAC_TX0, 15);
+    Interrupt_setPriority(INT_EMAC_RX0, 15);
+//    Interrupt_setPriority(INT_EMAC, 15);
+
+    //
+    //Enable the default interrupt handlers
+    //
+    Interrupt_enable(INT_EMAC_TX0);
+    Interrupt_enable(INT_EMAC_RX0);
+//    Interrupt_enable(INT_EMAC);
+
+    if(pInitCfg == NULL){
+        xReturn = pdFAIL;
+    }else{
+        NetworkEndPoint_t* pxEndPoint = FreeRTOS_FirstEndPoint(pxInterface);
+        uint8_t MACAddress[ipMAC_ADDRESS_LENGTH_BYTES] = pxEndPoint->xMACAddress.ucBytes;
+        uint32_t MACAddressLow = (pxEndPoint->xMACAddress.ucBytes[0]) | (pxEndPoint->xMACAddress.ucBytes[1] << 8) | (pxEndPoint->xMACAddress.ucBytes[2] << 16) | (pxEndPoint->xMACAddress.ucBytes[3]<<24);
+        uint32_t MACAddressHigh = (pxEndPoint->xMACAddress.ucBytes[4]) | (pxEndPoint->xMACAddress.ucBytes[5] << 8);
+        Ethernet_setMACAddr(EMAC_BASE, 0, MACAddressHigh, MACAddressLow, ETHERNET_CHANNEL_0);
+        xReturn = pdPASS;
+    }
+
+    if(xEMACTaskHandle == NULL){
+        xTaskCreate(prvEMACDeferredInterruptHandlerTask, "EMACInt", configMINIMAL_STACK_SIZE, NULL, configMAX_PRIORITIES - 1, &xEMACTaskHandle);
+        configASSERT(xEMACTaskHandle);
+        pxMyInterface = pxInterface;
+    }
+
+    return xReturn;
+}
+
+static BaseType_t prvf28388d_NetworkInterfaceOutput(NetworkInterface_t* pxInterface, NetworkBufferDescriptor_t* const pxDescriptor, BaseType_t xReleaseAfterSend){
+    pxGlobalDescriptor = pxDescriptor;
+    
+    Ethernet_Pkt_Desc pktDesc;
+    
+    pktDesc.bufferLength = pxDescriptor->xDataLength;
+    pktDesc.dataOffset = 0;
+    pktDesc.dataBuffer = pxDescriptor->pucEthernetBuffer;
+    pktDesc.nextPacketDesc = 0;
+    pktDesc.flags = ETHERNET_PKT_FLAG_SOP | ETHERNET_PKT_FLAG_EOP | ETHERNET_PKT_FLAG_SA_INS | ETHERNET_PKT_FLAG_SAIC;
+    pktDesc.pktChannel = ETHERNET_DMA_CHANNEL_NUM_0;
+    pktDesc.pktLength = pxDescriptor->xDataLength;
+    pktDesc.validLength = pxDescriptor->xDataLength;
+    pktDesc.numPktFrags = 1;
+
+    uint32_t ret = Ethernet_sendPacket(emac_handle, &pktDesc);
+
+    if(ret == 0){
+        vReleaseNetworkBufferAndDescriptor(pxDescriptor);
+    }
+
+    if(xReleaseAfterSend != pdFALSE){
+        pxDescriptor->pucEthernetBuffer = NULL;
+        vReleaseNetworkBufferAndDescriptor(pxDescriptor);

```

Comment:
In the case when `if(ret == 0)`, you have already called `vReleaseNetworkBufferAndDescriptor()`. It would cause an erro to call it again.